File: Examples/intel_examples/examples_core_c/fftw3xc/source/sp_plan_r2r_1d.c
/*******************************************************************************
* Copyright 2011-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
! Content:
!       Example of using fftwf_plan_r2r_1d function.
!
!****************************************************************************/

#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <float.h>
#include "fftw3.h"
#include "fftw3_mkl.h"

static void init_redft00(float *x, int N, int H);
static void init_redft10(float *x, int N, int H);
static void init_redft01(float *x, int N, int H);
static void init_redft11(float *x, int N, int H);
static void init_rodft00(float *x, int N, int H);
static void init_rodft10(float *x, int N, int H);
static void init_rodft01(float *x, int N, int H);
static void init_rodft11(float *x, int N, int H);
static int verify(float *x, int N, int H);

int main(void)
{
    /* Size of 1D r2r transform, N>1 */
    int N = 8;

    /* Arbitrary harmonic used to verify FFT */
    int H = 1;

    /* FFTW plan handle */
    fftwf_plan r2r = 0;

    /* Pointer to input/output data */
    float *x = 0;

    /* Execution status */
    int status = 0;

    printf("Example sp_plan_r2r_1d\n");
    printf("1D real-to-real in-place transform\n");
    printf("Configuration parameters:\n");
    printf(" N = %d\n", N);
    printf(" H = %d\n", H);

    printf("Allocate data array x(%i)\n", N+1); /* extra element for MKl_RODFT00 kind */
    x  = fftwf_malloc(sizeof(float)*(N+1));
    if (0 == x) goto failed;

    printf("============ REDFT00 (DCT1) ============\n");

    printf("Create FFTW_REDFT00 plan for N=%i\n",N);
    r2r = fftwf_plan_r2r_1d(N, x, x, FFTW_REDFT00, FFTW_ESTIMATE);
    if (0 == r2r) goto failed;

    printf("Initialize input for REDFT00 transform\n");
    init_redft00(x, N, H);

    printf("Compute FFTW_REDFT00\n");
    fftwf_execute(r2r);

    printf("Verify the result of REDFT00\n");
    status = verify(x, N, H);
    if (0 != status) goto failed;

    printf("Destroy FFTW plan\n");
    fftwf_destroy_plan(r2r);

    printf("============ REDFT10 (DCT2) ============\n");

    printf("Create FFTW_REDFT10 plan for N=%i\n",N);
    r2r = fftwf_plan_r2r_1d(N, x, x, FFTW_REDFT10, FFTW_ESTIMATE);
    if (0 == r2r) goto failed;

    printf("Initialize input for REDFT10 transform\n");
    init_redft10(x, N, H);

    printf("Compute FFTW_REDFT10\n");
    fftwf_execute(r2r);

    printf("Verify the result of REDFT10\n");
    status = verify(x, N, H);
    if (0 != status) goto failed;

    printf("Destroy FFTW plan\n");
    fftwf_destroy_plan(r2r);

    printf("============ REDFT01 (DCT3) ============\n");

    printf("Create FFTW_REDFT01 plan for N=%i\n",N);
    r2r = fftwf_plan_r2r_1d(N, x, x, FFTW_REDFT01, FFTW_ESTIMATE);
    if (0 == r2r) goto failed;

    printf("Initialize input for REDFT01 transform\n");
    init_redft01(x, N, H);

    printf("Compute FFTW_REDFT01\n");
    fftwf_execute(r2r);

    printf("Verify the result of REDFT01\n");
    status = verify(x, N, H);
    if (0 != status) goto failed;

    printf("Destroy FFTW plan\n");
    fftwf_destroy_plan(r2r);

    printf("============ REDFT11 (DCT4) ============\n");

    printf("Create FFTW_REDFT11 plan for N=%i\n",N);
    r2r = fftwf_plan_r2r_1d(N, x, x, FFTW_REDFT11, FFTW_ESTIMATE);
    if (0 == r2r) goto failed;

    printf("Initialize input for REDFT11 transform\n");
    init_redft11(x, N, H);

    printf("Compute FFTW_REDFT11\n");
    fftwf_execute(r2r);

    printf("Verify the result of REDFT11\n");
    status = verify(x, N, H);
    if (0 != status) goto failed;

    printf("Destroy FFTW plan\n");
    fftwf_destroy_plan(r2r);

    printf("============ RODFT00 (DST1) ============\n");

    printf("Create FFTW_RODFT00 plan for N=%i\n",N);
    r2r = fftwf_plan_r2r_1d(N, x, x, FFTW_RODFT00, FFTW_ESTIMATE);
    if (0 == r2r) goto failed;

    printf("Initialize input for RODFT00 transform\n");
    init_rodft00(x, N, H);

    printf("Compute FFTW_RODFT00\n");
    fftwf_execute(r2r);

    printf("Verify the result of RODFT00\n");
    status = verify(x, N, H);
    if (0 != status) goto failed;

    printf("Destroy FFTW plan\n");
    fftwf_destroy_plan(r2r);

    printf("============ RODFT10 (DST2) ============\n");

    printf("Create FFTW_RODFT10 plan for N=%i\n",N);
    r2r = fftwf_plan_r2r_1d(N, x, x, FFTW_RODFT10, FFTW_ESTIMATE);
    if (0 == r2r) goto failed;

    printf("Initialize input for RODFT10 transform\n");
    init_rodft10(x, N, H);

    printf("Compute FFTW_RODFT10\n");
    fftwf_execute(r2r);

    printf("Verify the result of RODFT10\n");
    status = verify(x, N, H);
    if (0 != status) goto failed;

    printf("Destroy FFTW plan\n");
    fftwf_destroy_plan(r2r);

    printf("============ RODFT01 (DST3) ============\n");

    printf("Create FFTW_RODFT01 plan for N=%i\n",N);
    r2r = fftwf_plan_r2r_1d(N, x, x, FFTW_RODFT01, FFTW_ESTIMATE);
    if (0 == r2r) goto failed;

    printf("Initialize input for RODFT01 transform\n");
    init_rodft01(x, N, H);

    printf("Compute FFTW_RODFT01\n");
    fftwf_execute(r2r);

    printf("Verify the result of RODFT01\n");
    status = verify(x, N, H);
    if (0 != status) goto failed;

    printf("Destroy FFTW plan\n");
    fftwf_destroy_plan(r2r);

    printf("============ RODFT11 (DST4) ============\n");

    printf("Create FFTW_RODFT11 plan for N=%i\n",N);
    r2r = fftwf_plan_r2r_1d(N, x, x, FFTW_RODFT11, FFTW_ESTIMATE);
    if (0 == r2r) goto failed;

    printf("Initialize input for RODFT11 transform\n");
    init_rodft11(x, N, H);

    printf("Compute FFTW_RODFT11\n");
    fftwf_execute(r2r);

    printf("Verify the result of RODFT11\n");
    status = verify(x, N, H);
    if (0 != status) goto failed;

    printf("Destroy FFTW plan\n");
    fftwf_destroy_plan(r2r);

    printf("============ MKL_RODFT00 (DST1) ============\n");

    printf("Create MKL_RODFT00 plan for N=%i\n",N);
    r2r = fftwf_plan_r2r_1d(N, x, x, (fftwf_r2r_kind)MKL_RODFT00, FFTW_ESTIMATE);
    if (0 == r2r) goto failed;

    printf("Initialize input for RODFT00 transform\n");
    init_rodft00(x+1, N, H);

    printf("Compute RODFT00 using new-array execute function for variety\n");
    fftwf_execute_r2r(r2r, x, x);

    printf("Verify the result of RODFT00\n");
    status = verify(x+1, N, H);
    if (0 != status) goto failed;

 cleanup:

    printf("Destroy FFTW plan\n");
    fftwf_destroy_plan(r2r);

    printf("Free data array\n");
    fftwf_free(x);

    printf("TEST %s\n",0==status ? "PASSED" : "FAILED");
    return status;

 failed:
    printf(" ERROR\n");
    status = 1;
    goto cleanup;
}

/* Compute (K*L)%M accurately */
static float moda(int K, int L, int M)
{
    return (float)(((long long)K * L) % M);
}

/* Compute K%L reduced to 0..L-1 */
static int modulo(int K, int L)
{
    return (K % L + L) % L;
}

/* Initialize array x(N) to produce unit peak at H */
static void init_redft00(float *x, int N, int H)
{
    float PI = 3.14159265358979323846f, factor;
    int n, j, NL;

    j = modulo(H,N);
    NL = 2*(N-1); /* logical size of transform */
    factor = ( j==0 || j==N-1 ) ? 1.0f : 2.0f;
    for (n = 0; n < N; n++)
    {
        x[n] = factor * cosf( 2*PI * moda(n,j,NL) / NL ) / NL;
    }
}

/* Initialize array x(N) to produce unit peak at H */
static void init_redft10(float *x, int N, int H)
{
    float PI = 3.14159265358979323846f, factor;
    int n, j, NL;

    j = modulo(H,N);
    NL = 2*N; /* logical size of transform */
    factor = ( j==0 ) ? 1.0f : 2.0f;
    for (n = 0; n < N; n++)
    {
        x[n] = factor * cosf( PI * moda(2*n+1,j,2*NL) / NL ) / NL;
    }
}

/* Initialize array x(N) to produce unit peak at H */
static void init_redft01(float *x, int N, int H)
{
    float PI = 3.14159265358979323846f, factor;
    int n, j, NL;

    j = modulo(H,N);
    NL = 2*N; /* logical size of transform */
    factor = 2;
    for (n = 0; n < N; n++)
    {
        x[n] = factor * cosf( PI * moda(n,2*j+1,2*NL) / NL ) / NL;
    }
}

/* Initialize array x(N) to produce unit peak at H */
static void init_redft11(float *x, int N, int H)
{
    float PI = 3.14159265358979323846f, factor;
    int n, j, NL;

    j = modulo(H,N);
    NL = 2*N; /* logical size of transform */
    factor = 2;
    for (n = 0; n < N; n++)
    {
        x[n] = factor * cosf( PI * moda(2*n+1,2*j+1,4*NL) / (2*NL) ) / NL;
    }
}

/* Initialize array x(N) to produce unit peak at H */
static void init_rodft00(float *x, int N, int H)
{
    float PI = 3.14159265358979323846f, factor;
    int n, j, NL;

    j = modulo(H,N);
    NL = 2*(N+1); /* logical size of transform */
    factor = 2;
    for (n = 0; n < N; n++)
    {
        x[n] = factor * sinf( 2*PI * moda(n+1,j+1,NL) / NL ) / NL;
    }
}

static void init_rodft10(float *x, int N, int H)
{
    float PI = 3.14159265358979323846f, factor;
    int n, j, NL;

    j = modulo(H,N);
    NL = 2*N; /* logical size of transform */
    factor = (j==N-1) ? 1.0f : 2.0f;
    for (n = 0; n < N; n++)
    {
        x[n] = factor * sinf( PI * moda(2*n+1,j+1,2*NL) / NL ) / NL;
    }
}

static void init_rodft01(float *x, int N, int H)
{
    float PI = 3.14159265358979323846f, factor;
    int n, j, NL;

    j = modulo(H,N);
    NL = 2*N; /* logical size of transform */
    factor = 2;
    for (n = 0; n < N; n++)
    {
        x[n] = factor * sinf( PI * moda(n+1,2*j+1,2*NL) / NL ) / NL;
    }
}

static void init_rodft11(float *x, int N, int H)
{
    float PI = 3.14159265358979323846f, factor;
    int n, j, NL;

    j = modulo(H,N);
    NL = 2*N; /* logical size of transform */
    factor = 2;
    for (n = 0; n < N; n++)
    {
        x[n] = factor * sinf( PI * moda(2*n+1,2*j+1,4*NL) / (2*NL) ) / NL;
    }
}

/* Verify that x has unit peak at H */
static int verify(float *x, int N, int H)
{
    float err, errthr, maxerr;
    int n;

    /*
     * Note, this simple error bound doesn't take into account error of
     * input data
     */
    errthr = 2.5f * logf( (float)N ) / logf(2.0f) * FLT_EPSILON;
    printf(" Check if err is below errthr %.3g\n", errthr);

    maxerr = 0;
    for (n = 0; n < N; n++)
    {
        float re_exp, re_got;

        re_exp = ((n-H)%N == 0 ? 1 : 0);
        re_got = x[n];

        err  = fabsf(re_got - re_exp);
        if (err > maxerr) maxerr = err;
        if (!(err < errthr))
        {
            printf(" x[%i]: ",n);
            printf(" expected %.7g, ",re_exp);
            printf(" got %.7g, ",re_got);
            printf(" err %.3g\n", err);
            printf(" Verification FAILED\n");
            return 1;
        }
    }
    printf(" Verified,  maximum error was %.3g\n", maxerr);
    return 0;
}

File: Examples/intel_examples/examples_core_f/fftw3xf/source/sp_plan_r2r_1d.f90
!===============================================================================
! Copyright 2011-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

! Content:
!       Example of using sfftw_plan_r2r_1d function.
!
!*****************************************************************************

program sp_plan_r2r_1d

  include 'fftw3.f'
  include 'fftw3_mkl.f'

  ! Physical size of 1D r2r transform, N>1
  integer, parameter :: N = 7

  ! Arbitrary harmonic used to verify FFT
  integer, parameter :: H = 1

  ! Working precision is single precision
  integer, parameter :: WP = selected_real_kind(6,37)

  ! Execution status
  integer :: status = 0

  ! Data array
  real(WP), allocatable :: x(:)

  ! FFTW plan
  integer*8 :: r2r = 0

  integer i

  print *,"Example sp_plan_r2r_1d"
  print *,"1D real-to-real transforms"
  print '("  N = "I0)', N
  print '("  H = "I0)', H

  print *,"Allocate data array"
  ! Extra element is needed for MKL_RODFT00 kind
  allocate ( x(N+1), STAT = status)
  if (0 /= status) goto 999

  print *,"========== REDFT00 (DCT1) ========="

  print '(" Create FFTW_REDFT00 plan for N="I0)', N
  call sfftw_plan_r2r_1d(r2r, N, x, x, FFTW_REDFT00, FFTW_ESTIMATE)
  if (0 == r2r) goto 999

  print '(" Initialize data for REDFT00, H="I0)', H
  call init_redft00(x, N, H)

  print *,"Compute FFTW_REDFT00 transform"
  !print '("X="10(F10.4:","))',x(1:N)
  call sfftw_execute(r2r)
  !print '("Y="10(F10.4:","))',x(1:N)

  print *,"Verify the result of FFTW_REDFT00 transform"
  status = verificate(x, N, H)
  if (0 /= status) goto 999

  print *,"Destroy FFTW plan"
  call sfftw_destroy_plan(r2r)

  print *,"========== REDFT10 (DCT2) ========="

  print '(" Create FFTW_REDFT10 plan for N="I0)', N
  call sfftw_plan_r2r_1d(r2r, N, x, x, FFTW_REDFT10, FFTW_ESTIMATE)
  if (0 == r2r) goto 999

  print '(" Initialize data for REDFT10, H="I0)', H
  call init_redft10(x, N, H)

  print *,"Compute FFTW_REDFT10 transform"
  !print '("X="10(F10.4:","))',x(1:N)
  call sfftw_execute(r2r)
  !print '("Y="10(F10.4:","))',x(1:N)

  print *,"Verify the result of FFTW_REDFT10 transform"
  status = verificate(x, N, H)
  if (0 /= status) goto 999

  print *,"Destroy FFTW plan"
  call sfftw_destroy_plan(r2r)

  print *,"========== REDFT01 (DCT3) ========="

  print '(" Create FFTW_REDFT01 plan for N="I0)', N
  call sfftw_plan_r2r_1d(r2r, N, x, x, FFTW_REDFT01, FFTW_ESTIMATE)
  if (0 == r2r) goto 999

  print '(" Initialize data for REDFT01, H="I0)', H
  call init_redft01(x, N, H)

  print *,"Compute FFTW_REDFT01 transform"
  !print '("X="10(F10.4:","))',x(1:N)
  call sfftw_execute(r2r)
  !print '("Y="10(F10.4:","))',x(1:N)

  print *,"Verify the result of FFTW_REDFT01 transform"
  status = verificate(x, N, H)
  if (0 /= status) goto 999

  print *,"Destroy FFTW plan"
  call sfftw_destroy_plan(r2r)

  print *,"========== REDFT11 (DCT4) ========="

  print '(" Create FFTW_REDFT11 plan for N="I0)', N
  call sfftw_plan_r2r_1d(r2r, N, x, x, FFTW_REDFT11, FFTW_ESTIMATE)
  if (0 == r2r) goto 999

  print '(" Initialize data for REDFT11, H="I0)', H
  call init_redft11(x, N, H)

  print *,"Compute FFTW_REDFT11 transform"
  !print '("X="10(F10.4:","))',x(1:N)
  call sfftw_execute(r2r)
  !print '("Y="10(F10.4:","))',x(1:N)

  print *,"Verify the result of FFTW_REDFT11 transform"
  status = verificate(x, N, H)
  if (0 /= status) goto 999

  print *,"Destroy FFTW plan"
  call sfftw_destroy_plan(r2r)

  print *,"========== RODFT00 (DST1) ========="

  print '(" Create FFTW_RODFT00 plan for N="I0)', N
  call sfftw_plan_r2r_1d(r2r, N, x, x, FFTW_RODFT00, FFTW_ESTIMATE)
  if (0 == r2r) goto 999

  print '(" Initialize data for RODFT00, H="I0)', H
  call init_rodft00(x, N, H)

  print *,"Compute FFTW_RODFT00 transform"
  !print '("X="10(F10.4:","))',x(1:N)
  call sfftw_execute(r2r)
  !print '("Y="10(F10.4:","))',x(1:N)

  print *,"Verify the result of FFTW_RODFT00 transform"
  status = verificate(x, N, H)
  if (0 /= status) goto 999

  print *,"Destroy FFTW plan"
  call sfftw_destroy_plan(r2r)

  print *,"========== RODFT10 (DST2) ========="

  print '(" Create FFTW_RODFT10 plan for N="I0)', N
  call sfftw_plan_r2r_1d(r2r, N, x, x, FFTW_RODFT10, FFTW_ESTIMATE)
  if (0 == r2r) goto 999

  print '(" Initialize data for RODFT10, H="I0)', H
  call init_rodft10(x, N, H)

  print *,"Compute FFTW_RODFT10 transform"
  !print '("X="10(F10.4:","))',x(1:N)
  call sfftw_execute(r2r)
  !print '("Y="10(F10.4:","))',x(1:N)

  print *,"Verify the result of FFTW_RODFT10 transform"
  status = verificate(x, N, H)
  if (0 /= status) goto 999

  print *,"Destroy FFTW plan"
  call sfftw_destroy_plan(r2r)

  print *,"========== RODFT01 (DST3) ========="

  print '(" Create FFTW_RODFT01 plan for N="I0)', N
  call sfftw_plan_r2r_1d(r2r, N, x, x, FFTW_RODFT01, FFTW_ESTIMATE)
  if (0 == r2r) goto 999

  print '(" Initialize data for RODFT01, H="I0)', H
  call init_rodft01(x, N, H)

  print *,"Compute FFTW_RODFT01 transform"
  !print '("X="10(F10.4:","))',x(1:N)
  call sfftw_execute(r2r)
  !print '("Y="10(F10.4:","))',x(1:N)

  print *,"Verify the result of FFTW_RODFT01 transform"
  status = verificate(x, N, H)
  if (0 /= status) goto 999

  print *,"Destroy FFTW plan"
  call sfftw_destroy_plan(r2r)

  print *,"========== RODFT11 (DST4) ========="

  print '(" Create FFTW_RODFT11 plan for N="I0)', N
  call sfftw_plan_r2r_1d(r2r, N, x, x, FFTW_RODFT11, FFTW_ESTIMATE)
  if (0 == r2r) goto 999

  print '(" Initialize data for RODFT11, H="I0)', H
  call init_rodft11(x, N, H)

  print *,"Compute FFTW_RODFT11 transform"
  !print '("X="10(F10.4:","))',x(1:N)
  call sfftw_execute(r2r)
  !print '("Y="10(F10.4:","))',x(1:N)

  print *,"Verify the result of FFTW_RODFT11 transform"
  status = verificate(x, N, H)
  if (0 /= status) goto 999

  print *,"Destroy FFTW plan"
  call sfftw_destroy_plan(r2r)

  print *,"========== MKL_RODFT00 (DST1) ========="

  print '(" Create MKL_RODFT00 plan for N="I0)', N
  call sfftw_plan_r2r_1d(r2r, N, x, x, MKL_RODFT00, FFTW_ESTIMATE)
  if (0 == r2r) goto 999

  print '(" Initialize data for MKL_RODFT00, H="I0)', H
  call init_rodft00(x, N, H)
  ! Unlike FFTW_RODFT00, MKL_RODFT00 requires extra zero element at x(1)
  ! Some compilers cannot do this well: x(2:1+N) = x(1:N)
  do i = N, 1, -1
    x(i+1) = x(i)
  end do
  x(1) = 0.0_WP

  print *,"Compute MKL_RODFT00 transform"
  !print '("X="10(F10.4:","))',x(1:1+N)
  call sfftw_execute(r2r)
  !print '("Y="10(F10.4:","))',x(1:1+N)

  print *,"Verify the result of MKL_RODFT00 transform"
  ! Ignore the extra zero element at x(1)
  x(1:N) = x(2:1+N)
  status = verificate(x, N, H)
  if (0 /= status) goto 999

  print *,"Destroy FFTW plan"
  call sfftw_destroy_plan(r2r)

  print *,"Deallocate data array"
  deallocate(x)

100 continue

  if (status == 0) then
    print *, "TEST PASSED"
    call exit(0)
  else
    print *, "TEST FAILED"
    call exit(1)
  end if

999 print '("  Error, status = ",I0)', status
  status = 1
  goto 100

contains

  ! Compute mod(K*L,M) accurately
  real(WP) function amoda(k,l,m)
    integer, intent(in) :: k,l,m
    integer*8 :: k8
    k8 = k
    amoda = real( mod(k8*l,m), WP )
  end function amoda

  ! Initialize x(:) to be inverse of (n==H)
  subroutine init_redft00(x, N, H)
    integer N, H
    real(WP) :: x(:)
    real(WP), parameter:: M_PI = 3.14159265358979323846_WP
    integer j, k, NL

    j = modulo(H,N)
    NL = 2*(N-1) ! logical size of transform
    do k = 1, N
      x(k) = 2.0 * cos( M_PI * amoda(j,k-1,NL) / real((N-1),WP) )
    end do
    if ( j==0 .or. j==N-1 ) x(:) = x(:) / 2.0
    x(:) = x(:) / real(NL,WP)
  end subroutine init_redft00

  ! Initialize x(:) to be inverse of (n==H)
  subroutine init_redft10(x, N, H)
    integer N, H
    real(WP) :: x(:)
    real(WP), parameter:: M_PI = 3.14159265358979323846_WP
    integer j, k, NL

    j = modulo(H,N)
    NL = 2*N ! logical size of transform
    do k = 1, N
      x(k) = 2.0 * cos( M_PI * amoda(j,2*k-1,2*NL) / real((2*N),WP) )
    end do
    if (j==0) x(:) = x(:) / 2.0
    x(:) = x(:) / real(NL,WP)
  end subroutine init_redft10

  ! Initialize x(:) to be inverse of (n==H)
  subroutine init_redft01(x, N, H)
    integer N, H
    real(WP) :: x(:)
    real(WP), parameter:: M_PI = 3.14159265358979323846_WP
    integer j, k, NL

    j = modulo(H,N)
    NL = 2*N ! logical size of transform
    do k = 1, N
      x(k) = 2.0 * cos( M_PI * amoda(2*j+1,k-1,2*NL) / real((2*N),WP) )
    end do
    x(:) = x(:) / real(NL,WP)
  end subroutine init_redft01

  ! Initialize x(:) to be inverse of (n==H)
  subroutine init_redft11(x, N, H)
    integer N, H
    real(WP) :: x(:)
    real(WP), parameter:: M_PI = 3.14159265358979323846_WP
    integer j, k, NL

    j = modulo(H,N)
    NL = 2*N ! logical size of transform
    do k = 1, N
      x(k) = 2.0 * cos( M_PI * amoda(2*j+1,2*k-1,4*NL) / real((4*N),WP) )
    end do
    x(:) = x(:) / real(NL,WP)
  end subroutine init_redft11

  ! Initialize x(:) to be inverse of (n==H)
  subroutine init_rodft00(x, N, H)
    integer N, H
    real(WP) :: x(:)
    real(WP), parameter:: M_PI = 3.14159265358979323846_WP
    integer j, k, NL

    j = modulo(H,N)
    NL = 2*(N+1) ! logical size of transform
    do k = 1, N
      x(k) = 2.0 * sin( M_PI * amoda(j+1,k,NL) / real((N+1),WP) )
    end do
    x(:) = x(:) / real(NL,WP)
  end subroutine init_rodft00

  ! Initialize x(:) to be inverse of (n==H)
  subroutine init_rodft10(x, N, H)
    integer N, H
    real(WP) :: x(:)
    real(WP), parameter:: M_PI = 3.14159265358979323846_WP
    integer j, k, NL

    j = modulo(H,N)
    NL = 2*N ! logical size of transform
    do k = 1, N
      x(k) = 2.0 * sin( M_PI * amoda(j+1,2*k-1,2*NL) / real((2*N),WP) )
    end do
    if (j==N-1) x(:) = x(:) / 2.0
    x(:) = x(:) / real(NL,WP)
  end subroutine init_rodft10

  ! Initialize x(:) to be inverse of (n==H)
  subroutine init_rodft01(x, N, H)
    integer N, H
    real(WP) :: x(:)
    real(WP), parameter:: M_PI = 3.14159265358979323846_WP
    integer j, k, NL

    j = modulo(H,N)
    NL = 2*N ! logical size of transform
    do k = 1, N
      x(k) = 2.0 * sin( M_PI * amoda(2*j+1,k,2*NL) / real((2*N),WP) )
    end do
    x(:) = x(:) / real(NL,WP)
  end subroutine init_rodft01

  ! Initialize x(:) to be inverse of (n==H)
  subroutine init_rodft11(x, N, H)
    integer N, H
    real(WP) :: x(:)
    real(WP), parameter:: M_PI = 3.14159265358979323846_WP
    integer j, k, NL

    j = modulo(H,N)
    NL = 2*N ! logical size of transform
    do k = 1, N
      x(k) = 2.0 * sin( M_PI * amoda(2*j+1,2*k-1,4*NL) / real((4*N),WP) )
    end do
    x(:) = x(:) / real(NL,WP)
  end subroutine init_rodft11

  ! Verify that x(:) is unit peak at x(modulo(H,N))
  integer function verificate(x, N, H)
    integer N, H
    real(WP) :: x(:)

    integer j,k
    real(WP) err, errthr, maxerr, res_exp
    real(WP), parameter:: M_PI = 3.14159265358979323846_WP

    ! Note, this simple error bound doesn't take into account error of
    ! input data
    errthr = 2.5 * log(real(N, WP)) / log(2.0_WP) * EPSILON(1.0_WP)
    print '("  Check if err is below errthr " G10.3)', errthr
    j = modulo(H,N) + 1
    maxerr = 0.0_WP
    do k = 1,N
      if (k==j) then
        res_exp = 1.0_WP
      else
        res_exp = 0.0_WP
      end if
      err = abs(x(k) - res_exp)
      maxerr = max(err,maxerr)
      if (.not.(err < errthr)) then
        print '("  x("I0"):"$)', k
        print '(" expected "G14.7","$)', res_exp
        print '(" got "G14.7","$)', x(k)
        print '(" err "G10.3)', err
        print *,"  Verification FAILED"
        verificate = 1
        return
      end if
    end do
    print '("  Verified,  maximum error was " G10.3)', maxerr
    verificate = 0
  end function verificate

end program sp_plan_r2r_1d
